src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -111,6 +111,249 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+        finfo->ep_attr->tx_ctx_cnt = MPIDI_Global.max_rma_sep_tx_cnt;
+
+        MPIDI_OFI_CALL_RETURN(fi_scalable_ep
+                              (MPIDI_Global.domain, finfo, &MPIDI_Global.rma_sep, NULL), ret);
+        if (ret < 0) {
+            MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE, "Failed to create scalable endpoint.\n");
+            MPIDI_Global.rma_sep = NULL;
+            mpi_errno = MPIDI_OFI_ENAVAIL;
+            goto fn_fail;
+        }
+
+        MPIDI_OFI_CALL_RETURN(fi_scalable_ep_bind(MPIDI_Global.rma_sep, &(MPIDI_Global.av->fid), 0),
+                              ret);
+        if (ret < 0) {
+            MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                        "Failed to bind scalable endpoint to address vector.\n");
+            MPIDI_Global.rma_sep = NULL;
+            mpi_errno = MPIDI_OFI_ENAVAIL;
+            goto fn_fail;
+        }
+
+        /* Allocate and initilize tx index array. */
+        utarray_new(MPIDI_Global.rma_sep_idx_array, &ut_int_icd, MPL_MEM_RMA);
+        for (i = 0; i < MPIDI_Global.max_rma_sep_tx_cnt; i++) {
+            utarray_push_back(MPIDI_Global.rma_sep_idx_array, &i, MPL_MEM_RMA);
+        }
+    }
+    /* Set per window transmit attributes. */
+    MPIDI_OFI_set_rma_fi_info(win, finfo);
+    /* Get available transmit context index. */
+    int *index = (int *) utarray_back(MPIDI_Global.rma_sep_idx_array);
+    if (index == NULL) {
+        mpi_errno = MPIDI_OFI_ENAVAIL;
+        goto fn_fail;
+    }
+    /* Retrieve transmit context on scalable EP. */
+    MPIDI_OFI_CALL_RETURN(fi_tx_context
+                          (MPIDI_Global.rma_sep, *index, finfo->tx_attr, &(MPIDI_OFI_WIN(win).ep),
+                           NULL), ret);
+    if (ret < 0) {
+        MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    "Failed to retrieve transmit context from scalable endpoint.\n");
+        mpi_errno = MPIDI_OFI_ENAVAIL;
+        goto fn_fail;
+    }
+
+    MPIDI_OFI_WIN(win).sep_tx_idx = *index;
+    /* Pop this index out of reserving array. */
+    utarray_pop_back(MPIDI_Global.rma_sep_idx_array);
+
+    MPIDI_OFI_CALL_RETURN(fi_ep_bind(MPIDI_OFI_WIN(win).ep,
+                                     &MPIDI_Global.ctx[0].cq->fid,
+                                     FI_TRANSMIT | FI_SELECTIVE_COMPLETION), ret);
+    if (ret < 0) {
+        MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    "Failed to bind endpoint to completion queue.\n");
+        mpi_errno = MPIDI_OFI_ENAVAIL;
+        goto fn_fail;
+    }
+
+    /* Check whether per-window counter is used. */
+    if (MPIDI_OFI_win_set_per_win_sync(win) == MPI_SUCCESS) {
+        MPIDI_OFI_CALL_RETURN(fi_enable(MPIDI_OFI_WIN(win).ep), ret);
+        if (ret < 0) {
+            MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE, "Failed to activate endpoint.\n");
+            mpi_errno = MPIDI_OFI_EPERROR;
+            goto fn_fail;
```

Comment:
1. It is a little risky that you define `MPIDI_OFI_EPERROR` and `MPIDI_OFI_ENAVAIL` in OFI and assume they are not equal to `MPI_SUCCESS`. Perhaps it would be better if you just do not use the MPI error code (including `MPI_SUCCESS`) in `MPIDI_OFI_win_init_sep`.

2. Could you please also change the error returning mechanism in `MPIDI_OFI_win_init_stx` ? They need to be consistent.

Reply:
I'm thinking about using `MPIDI_OFI_SUCCESS` as return value for all those endpoint initialization function, to replace `MPI_SUCCESS`.

Reply:
@minsii I don't want to disturb the discussion too much, but can't we assume negative values won't conflict with MPI/MPICH error codes? Actually we are already doing similar thing here: https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/netmod/ofi/ofi_send.h#L79

Reply:
@hajimefu : You are right that all MPI error codes are positive numbers as required in standard. So the negative error codes will not be conflict with any MPI error code.
Section 8.4 says: ```0 = MPI_SUCCESS < MPI_ERR...```

Perhaps we should at least add a comment on top of the function similar to `ofi_send.h`.

Reply:
I can add similar comments.